### DISCUSSION TITLE
Scroll lots of lines in a performant way

### DISCUSSION THREAD

**Discussion by matthesoundman** [ArthurSonzogni/FTXUI#304] [Q&A] [answered]
Hello,

I am attempting to simply read in some large log files and merge them. Then show the merged log lines with the ability to scroll. I have plans to colorize certain parts of the log entry etc. There is more than 150,000 lines merged and parsed.

I am using the `Menu()` as a scroller. It has nice highlighting so I thought this might suffice.

However, right now I am getting poor performance. The highlight and scroll has a large delay making the app unusable when going through the lines.

```cpp
int main(int argc, const char *argv[]) {
  if (argc != 3) {
    LOG_ERROR("Incorrect arguments provided");
    return EXIT_FAILURE;
  }
  
  std::vector<std::string> lines;
  readLogFiles(argv[1], argv[2], lines);
  
  int selected {0};
  auto menu {Menu(&lines, &selected)};
  
  auto scroller {Renderer(menu, [&]() {
    return menu->Render() | frame | border;
  })};
  
  auto screen = ScreenInteractive::Fullscreen();
  screen.Loop(scroller);
  
  return EXIT_SUCCESS;
```

Any tips or suggestions on how to make the refresh faster for hundred thousands of lines?

Thank you kindly.

**Comment by ArthurSonzogni** [comment, marked as answer]
Hi @matthesoundman 

Yes, you can reimplement the menu yourself and draw only a shorter range?

Here, I am rewriting the Render() method from the normal Menu.

For instance:
```cpp
int selected_line = 0;
std::vector<std::string> lines;
auto menu = Menu(&selected_line, &lines);

auto my_custom_menu = Renderer(menu, [&] {
  int begin = std::max(0, selected_lines - 10);
  int end = std::min(lines.size(), selected_lines + 10)
  Elements elements;
  for(int i = begin; i<end; ++i) {
    Element element = text(lines[i]);
    if (i == selected_line)
      element = element | inverted | selected;
    element.push_back(element);
  }
  return vbox(std::move(elements)) | vscroll_indicator | frame | border;
});
```

You may also want to create your own Component by deriving from ComponentBase.

**Reply by matthesoundman** [reply]
Thanks. I will give it a try and see how it goes.

**Reply by matthesoundman** [reply]
Hi @ArthurSonzogni,
This is a lot better. Thank you for the help.

Do you have tips on how I could adjust the max and min on the fly based on the window size?

**Reply by ArthurSonzogni** [reply]
From:
```
include/ftxui/screen/terminal.hpp
```
You can use:
```cpp
Terminal::Size().dimy;
```

**Reply by matthesoundman** [reply]
Awesome! Thank you very much. It works well now.
